tests/contrib/patch.py

```diff
@@ -280,15 +304,14 @@ def test_patch_unpatch_patch(self):
                 self.assert_wrapped(redis.client.BasePipeline.execute)
                 self.assert_wrapped(redis.client.BasePipeline.immediate_execute_command)
             """
-            for module_name, integration_name, unpatch in self.MODULES:
-                self.assert_module_not_imported(module_name)
-                self.patch(**{integration_name: True})
-                unpatch()
-                self.patch(**{integration_name: True})
-                module = importlib.import_module(module_name)
-                self.assert_patched(module)
-
-        @require_modules
+            self.assert_module_not_imported(self.__module_name__)
+            self.patch(**{self.__integration_name__: True})
+            self.__unpatch_func__()
```

Comment:
Should we do:

```
patch()
assert_patched()
unpatch()
assert_not_patched()
patch()
assert_patched()
```

?

Reply:
Since we need to import the module to check if it is patched, I opted to test the logic like this _before_ the module is imported. Technically with each of these tests we could have a permutation where we import at a different point. For example we could have:

```
import module
patch()
assert_patched(module)
unpatch()
assert_not_patched(module)
patch()
assert_patched(module)
```
and
```
patch()
import module
assert_patched(module)
unpatch()
assert_not_patched(module)
patch()
assert_patched(module)
```
and 
```
patch()
unpatch()
import module
assert_not_patched(module)
patch()
assert_patched(module)
```
and
```
patch()
unpatch()
patch()
import module
assert_patched(module)
```

I think it does make sense to have a test case for the way the test is currently written, but it also makes sense to have a test case for the snippet you provided as well.

I tried to keep the test cases minimal since when starting it seemed like we'd need to write these for each integration. But now that we can do this automatically I think we can add test cases for each of the above and more. What do you think?

Reply:
Cool. Yeah, I think especially since we are centralizing/deduping for all of our patch tests, then lets go crazy, better to overtest here than under test, since it is so easy to do (since it is all i one place).

Lets go with a solid minimal set of tests that at least gets us close to the types of other tests we have today, then we can always add more later. Even if you think of a list of tests to add, add a `todo` or something, that is ok

Reply:
as we add tests for each integration, we can also take the time to expand our test cases.